sw/host/opentitanlib/src/transport/proxy/spi.rs

```diff
@@ -0,0 +1,173 @@
+            )),
+        }
+    }
+
+    fn set_transfer_mode(&self, mode: TransferMode) -> Result<()> {
+        match self.execute_command(SpiRequest::SetTransferMode { mode })? {
+            SpiResponse::SetTransferMode => Ok(()),
+            _ => Err(TransportError::ProxyCommunicationError(
+                "Unexpected reply".to_string(),
+            )),
+        }
+    }
+
+    fn get_bits_per_word(&self) -> Result<u32> {
+        match self.execute_command(SpiRequest::GetBitsPerWord)? {
+            SpiResponse::GetBitsPerWord { bits_per_word } => Ok(bits_per_word),
+            _ => Err(TransportError::ProxyCommunicationError(
+                "Unexpected reply".to_string(),
+            )),
+        }
+    }
+    fn set_bits_per_word(&self, bits_per_word: u32) -> Result<()> {
+        match self.execute_command(SpiRequest::SetBitsPerWord { bits_per_word })? {
+            SpiResponse::SetBitsPerWord => Ok(()),
+            _ => Err(TransportError::ProxyCommunicationError(
+                "Unexpected reply".to_string(),
+            )),
+        }
+    }
+
+    fn get_max_speed(&self) -> Result<u32> {
+        match self.execute_command(SpiRequest::GetMaxSpeed)? {
+            SpiResponse::GetMaxSpeed { speed } => Ok(speed),
+            _ => Err(TransportError::ProxyCommunicationError(
+                "Unexpected reply".to_string(),
+            )),
+        }
+    }
+    fn set_max_speed(&self, value: u32) -> Result<()> {
+        match self.execute_command(SpiRequest::SetMaxSpeed { value })? {
+            SpiResponse::SetMaxSpeed => Ok(()),
+            _ => Err(TransportError::ProxyCommunicationError(
+                "Unexpected reply".to_string(),
+            )),
+        }
+    }
+
+    fn get_max_transfer_count(&self) -> Result<usize> {
+        match self.execute_command(SpiRequest::GetMaxTransferCount)? {
+            SpiResponse::GetMaxTransferCount { number } => Ok(number as usize),
+            _ => Err(TransportError::ProxyCommunicationError(
+                "Unexpected reply".to_string(),
+            )),
+        }
+    }
+
+    fn max_chunk_size(&self) -> Result<usize> {
+        match self.execute_command(SpiRequest::GetMaxChunkSize)? {
+            SpiResponse::GetMaxChunkSize { size } => Ok(size as usize),
+            _ => Err(TransportError::ProxyCommunicationError(
+                "Unexpected reply".to_string(),
+            )),
+        }
+    }
+
+    fn set_voltage(&self, voltage: Voltage) -> Result<()> {
+        match self.execute_command(SpiRequest::SetVoltage { voltage })? {
+            SpiResponse::SetVoltage => Ok(()),
+            _ => Err(TransportError::ProxyCommunicationError(
+                "Unexpected reply".to_string(),
+            )),
+        }
+    }
+
+    fn run_transaction(&self, transaction: &mut [Transfer]) -> Result<()> {
+        let mut req: Vec<SpiTransferRequest> = Vec::new();
+        for transfer in &*transaction {
+            // &* to treat as non-mutable in this loop
```

Comment:
Why is this necessary? It doesn't seem obvious to me.

Reply:
Without the ```&*```, I get this error:
```
105 |         let mut req: Vec<SpiTransferRequest> = Vec::new();
106 |         for transfer in transaction {
    |                         -----------
    |                         |
    |                         `transaction` moved due to this implicit call to `.into_iter()`
    |                         help: consider borrowing to avoid moving into the for loop: `&transaction`
...
129 |                     resp.len() == transaction.len(),
    |                                   ^^^^^^^^^^^^^^^^^ value borrowed here after move
```

If I add just a single ```&``` as suggested above, I instead get:
```
106 |         for transfer in &transaction {
    |                         ^^^^^^^^^^^^ `&&mut [io::spi::Transfer<'_, '_>]` is not an iterator
```

So that is why I ended up with the strangely looking ```&*``` to temporarily convert the mutable reference into an immutable reference.  (Maybe there is a more explicit way of doing the same, that I am not aware of.)

Reply:
This also works, may that is slightly better than `&*`:
```
for transfer in transaction.iter() {
```

Reply:
Neat error! Yeah iter() is probably nicer all around.

Reply:
Done.